### PR TITLE
feat: add merge readiness tooling

### PR DESCRIPTION
**Pull request by flyingrobots**
## Summary
- add merge-readiness reporting to the PR status helper
- down-rank stale historical CodeRabbit rate-limit comments when newer bot/check signals are green
- add a dedicated `npm run pr:merge-readiness` entrypoint and regression coverage

## Verification
- `npx vitest run --config vitest.config.ts scripts/pr-review-status.test.ts`
- `npm run typecheck:test`
- `npm run pr:review-status -- 43`
- `npm run pr:merge-readiness -- 43`
- `npm test`

## Summary by CodeRabbit

* **New Features**
  * Added a merge-readiness command/mode that summarizes PR merge gates, collapses reviewer signals to the latest non-automated review, honors review/merge state, down-ranks stale bot chatter (rate-limit and older signals), and returns a readiness exit status.

* **Tests**
  * Expanded coverage for review summarization, bot-status handling (stale vs active), rate-limit scenarios, pending/blocked/ready gating, and exit-code behavior.

* **Documentation**
  * Clarified handling of stale bot comments versus active bot runs and documented merge-readiness behavior.

### CODE REVIEW

File: scripts/pr-review-status.ts [hunk 1]
```diff
@@ -101,6 +141,194 @@ export function extractUnresolvedFindings(threads: readonly ReviewThreadNode[]):
...
+    reasons.push(`${input.unresolvedCount} unresolved review thread${input.unresolvedCount === 1 ? '' : 's'}`);
+  }
+
+  if (input.reviews.changesRequested > 0) {
+    reasons.push(`${input.reviews.changesRequested} change-request review${input.reviews.changesRequested === 1 ? '' : 's'}`);
+  }
+
+  if (input.reviews.total < input.minReviews) {
+    reasons.push(`needs at least ${input.minReviews} review${input.minReviews === 1 ? '' : 's'} (found ${input.reviews.total})`);
+  }
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply scripts/pr-review-status.ts:388]: Addressed in 1843d86. computeMergeReadiness() now blocks on GitHub reviewDecision values like REVIEW_REQUIRED and CHANGES_REQUESTED instead of relying only on raw review counts.

File: scripts/pr-review-status.ts [hunk 2]
```diff
@@ -101,6 +141,194 @@ export function extractUnresolvedFindings(threads: readonly ReviewThreadNode[]):
...
+    pendingReasons.push('CodeRabbit review still pending');
+  } else if (input.codeRabbit.state === 'rate_limited') {
+    pendingReasons.push('CodeRabbit is rate-limited');
+  }
+
+  if (pendingReasons.length > 0) {
+    return { status: 'pending', reasons: pendingReasons };
+  }
+
+  return { status: 'ready', reasons: [] };
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply scripts/pr-review-status.ts:414]: Addressed in 1843d86. computeMergeReadiness() now incorporates GitHub reviewDecision and mergeStateStatus directly in the early blocking logic, with regression coverage in scripts/pr-review-status.test.ts.

File: scripts/pr-review-status.ts [hunk 3]
```diff
@@ -101,6 +141,194 @@ export function extractUnresolvedFindings(threads: readonly ReviewThreadNode[]):
...
+  }
+
+  return {
+    total: reviews.length,
+    approvals: byState.APPROVED ?? 0,
+    changesRequested: byState.CHANGES_REQUESTED ?? 0,
+    comments: byState.COMMENTED ?? 0,
+    byState,
+  };
+}
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply scripts/pr-review-status.ts:227]: Addressed in 1843d86. summarizeReviews() now collapses to the latest non-automated review per reviewer, so repeated review history and bot comments no longer inflate the merge gate.

File: scripts/pr-review-status.ts [hunk 4]
```diff
@@ -101,6 +141,194 @@ export function extractUnresolvedFindings(threads: readonly ReviewThreadNode[]):
...
+    detail: 'missing',
+    latestKind: 'none',
+    staleRateLimitCount,
+    activeRateLimitCount,
+  };
+}
+
+export function computeMergeReadiness(input: {
+  readonly pr: Pick<PullRequestView, 'state' | 'isDraft'>;
+  readonly checks: CheckSummary;
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply scripts/pr-review-status.ts]: Addressed in 1843d86. computeMergeReadiness() now consumes mergeStateStatus and blocks on non-mergeable states, so conflicted or otherwise blocked PRs no longer report READY.

File: scripts/pr-review-status.ts [hunk 5]
```diff
@@ -211,6 +509,56 @@ function summarizeComment(body: string): string {
...
+): readonly { readonly at: string; readonly kind: CodeRabbitEventKind }[] {
+  const commentEvents = comments
+    .filter((comment) => comment.author?.login === 'coderabbitai')
+    .map((comment) => ({
+      at: comment.createdAt,
+      kind: classifyCodeRabbitBody(comment.body),
+    }));
+  const reviewEvents = reviews
+    .filter((review) => review.author?.login === 'coderabbitai')
+    .map((review) => ({
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply scripts/pr-review-status.ts]: Addressed in 7483f7d. collectCodeRabbitEvents() now matches both `coderabbitai` and `coderabbitai[bot]`, and the regression is covered in scripts/pr-review-status.test.ts.

File: docs/CODERABBIT_REVIEW_NOTES.md [hunk 6]
```diff
@@ -15,6 +15,11 @@ Track recurring friction in the PR feedback loop and concrete fixes (scripts/pro
 - **Nits can hide real regressions**: trivial findings occasionally surface useful edge cases (e.g., split-pane multi-width divider).
   Proposed improvement: keep the policy of fixing all severities, but auto-group by subsystem to batch test runs efficiently.
 
+### 2026-03-13
+- **Historical bot chatter still competes with live bot state**: older CodeRabbit comments can remain prominent even when a newer check is green or a new review is actively in progress.
+  Progress: `pr:review-status` already down-ranks stale historical `rate-limited` comments when a newer green CodeRabbit signal exists.
+  Proposed improvement: extend that logic across the full bot lifecycle (`pending`, `pass`, `rate-limited`, `no actionable comments`) so humans do not have to infer the current state manually from mixed historical comments.
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply docs/CODERABBIT_REVIEW_NOTES.md]: Addressed in 4673301. docs/CODERABBIT_REVIEW_NOTES.md now marks the shipped pending/pass behavior as implemented and narrows the remaining gap to older clean/actionable bot chatter.

File: scripts/pr-review-status.ts [hunk 7]
```diff
@@ -172,7 +472,16 @@ function main(): void {
     }
   }
 
-  process.exitCode = computeExitCode(summary, unresolved.length);
+  if (options.mergeReady && mergeReadiness.reasons.length > 0) {
+    process.stdout.write(`\nMerge blockers:\n`);
+    for (const reason of mergeReadiness.reasons) {
+      process.stdout.write(`- ${reason}\n`);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply scripts/pr-review-status.ts:522]: Addressed in 64da8cd. pr-review-status now uses a pending-specific heading via mergeReadinessHeading(), so pending states no longer print under "Merge blockers:".

File: scripts/pr-review-status.ts [hunk 8]
```diff
@@ -211,6 +520,124 @@ function summarizeComment(body: string): string {
...
+        }
+      }
+      reviews(first: 100) {
+        nodes {
+          body
+          submittedAt
+          state
+          author { login __typename }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply scripts/pr-review-status.ts:687]: Addressed in 64da8cd. The PR metadata query now requests pageInfo/totalCount for comments and reviews, and assertUntruncatedPullRequestData() fails fast if GitHub reports truncation so readiness is never computed from incomplete data.

File: scripts/pr-review-status.ts [hunk 9]
```diff
@@ -22,6 +26,19 @@ interface CheckEntry {
...
+  readonly body: string;
+  readonly createdAt: string;
+}
+
+interface PullRequestReview {
+  readonly author?: { readonly login: string; readonly __typename?: string } | null;
+  readonly body: string;
+  readonly submittedAt: string;
+  readonly state: string;
+}
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply scripts/pr-review-status.ts:40]: Addressed in 0438a45. PullRequestReview.submittedAt is now nullable, and both summarizeReviews() and collectCodeRabbitEvents() ignore draft/pending reviews until they have a submitted timestamp.

File: scripts/pr-review-status.ts [hunk 10]
```diff
@@ -61,6 +102,52 @@ export interface UnresolvedFinding {
...
+  }
+
+  if (input.reviews.pageInfo.hasNextPage) {
+    truncated.push(`reviews=${input.reviews.totalCount}`);
+  }
+
+  if (truncated.length > 0) {
+    throw new Error(`pull request metadata truncated; pagination required for ${truncated.join(', ')}`);
+  }
+}
```
[the remark this answers is not in the document]
Comment (flyingrobots) [reply scripts/pr-review-status.ts:153]: Addressed in 0438a45. fetchReviewThreads() now requests pageInfo/totalCount and routes through assertUntruncatedPullRequestData(), so unresolved-thread counts can no longer be computed from truncated review-thread data.
